src/v/coproc/router.cc

```diff
@@ -384,4 +372,109 @@ router::reader_cfg(model::offset start, model::offset end) {
       _abort_source);
 }
 
+ss::future<> router::process_additions() {
+    return ss::do_with(
+      std::move(_deferred_additions),
+      [this](absl::flat_hash_map<script_id, source<errc>>& dfa) {
+          return ss::do_for_each(
+            dfa.begin(),
+            dfa.end(),
+            [this](std::pair<const script_id, source<errc>>& element) {
+                auto& [id, src] = element;
+                // For now only support the 'latest' policy
+                if (!is_valid_ingestion_policy(src.tip)) {
+                    src.promise.set_value(errc::invalid_ingestion_policy);
+                    return ss::now();
+                }
+                auto logs = _api.local().log_mgr().get(src.tn);
+                if (logs.empty()) {
+                    src.promise.set_value(errc::topic_does_not_exist);
+                    return ss::now();
+                }
+
+                for (auto& [ntp, log] : logs) {
+                    auto found = _sources.find(ntp);
+                    if (found == _sources.end()) {
+                        auto ts = ss::make_lw_shared<topic_state>(topic_state{
+                          .log = log,
+                          .head = topic_offsets(),
+                          .scripts = {id}});
+                        _sources.emplace(ntp, std::move(ts));
+                    } else {
+                        found->second->scripts.emplace(id);
+                    }
+                    vlog(coproclog.info, "Inserted ntp {} id {}", ntp, id);
+                }
+                src.promise.set_value(errc::success);
+                return ss::now();
+            });
+      });
+}
+
+ss::future<> router::process_removals() {
+    return ss::do_with(
+      std::move(_deferred_removals),
```

Comment:
same here. std::exchange ?

Reply:
this is marked resolved without any change or follow up

Reply:
this has been resolved